Providers/Pocket/PocketSponsoredStoriesProvider.swift

```diff
@@ -0,0 +1,127 @@
+        }
+    }()
+
+    private var sponsoredFeedRequest: URLRequest? {
+        var request = Locale
+            .current
+            .regionCode?
+            .map { URLQueryItem(name: .country, value: $0) }
+            .map { endpoint.withQueryParams([$0]) }
+            .map {
+                URLRequest(
+                    url: $0,
+                    cachePolicy: .reloadIgnoringCacheData,
+                    timeoutInterval: 5
+                )
+            }
+
+        let body: [String: Any] = [
+            .version: 2,
+            .consumer_key: pocketKey,
+            .pocket_id: pocketId
+        ]
+
+        let bodyData = try? JSONSerialization.data(withJSONObject: body)
+        request?.httpBody = bodyData
+        request?.httpMethod = .post
+        return request
+    }
+
+    func fetchSponsoredStories(timestamp: Timestamp = Date.now(), completion: @escaping (SponsoredStoryResult) -> Void) {
+        guard let request = sponsoredFeedRequest else {
+            completion(.failure(Error.requestCreationFailure))
+            return
+        }
+
+        if let cachedData = findCachedData(for: request, timestamp: timestamp) {
+            decode(data: cachedData, completion: completion)
+        } else {
+            fetchSponsoredStories(request: request, completion: completion)
+        }
+    }
+
+    func fetchSponsoredStories(request: URLRequest, completion: @escaping (SponsoredStoryResult) -> Void) {
+        urlSession.dataTask(with: request) { [weak self] (data, response, error) in
+            guard let self = self else { return }
+            if let error = error {
+                self.browserLog.debug("An error occurred while fetching data: \(error)")
+                completion(.failure(Error.failure))
+                return
+            }
+
+            guard let response = validatedHTTPResponse(response, statusCode: 200..<300), let data = data, !data.isEmpty else {
+                self.browserLog.debug("Response isn't proper: \(response.debugDescription), with data \(String(describing: data))")
+                completion(.failure(Error.invalidHTTPResponse))
+                return
+            }
+
+            self.cache(response: response, for: request, with: data)
+            self.decode(data: data, completion: completion)
+        }.resume()
+    }
+
+    private func decode(data: Data, completion: @escaping (SponsoredStoryResult) -> Void) {
+        do {
+            let decodedResponse = try JSONDecoder().decode(PocketSponsoredRequest.self, from: data)
+            completion(.success(decodedResponse.spocs))
+        } catch {
+            self.browserLog.error("Unable to parse with error: \(error)")
+            completion(.failure(Error.decodingFailure))
+        }
+    }
+}
+
+extension PocketSponsoredStoriesProvider {
+    enum Error: Swift.Error {
+        case failure
+        case decodingFailure
+        case requestCreationFailure
+        case invalidHTTPResponse
+    }
```

Comment:
Yeah, this would make sense more as `PocketSponsoredStoriesProviderError` because you have an `Error.failure` somewhere else already.

Reply:
Similar with the explanation before, naming it as `PocketSponsoredStoriesProviderError` would not be a issue, but because it's nested, accessing it from the outside would be `PocketSponsoredStoriesProvider. PocketSponsoredStoriesProviderError` and it a bit repetitive.